File: CSharp/LeetCode/0200_0299/0283_MoveZeroes.cs
namespace LeetCode;

public class MoveZeroes
{
    public void MoveZeroesLinear(int[] nums)
    {
        int length = nums.Length;
        int[] newArray = new int[length];
        int newArrayPointer = 0;

        for (int i = 0; i < length; i++)
        {
            if (nums[i] != 0)
            {
                newArray[newArrayPointer] = nums[i];
                newArrayPointer++;
            }
        }

        for (int i = 0; i < length; i++)
        {
            nums[i] = newArray[i];
        }
    }

    public void MoveZeroesTwoPointer(int[] nums)
    {
        int length = nums.Length;
        int swapIndex = 0;

        for (int i = 0; i < length; i++)
        {
            if (nums[i] != 0)
            {
                (nums[swapIndex], nums[i]) = (nums[i], nums[swapIndex]);
                swapIndex++;
            }
        }
    }
}

File: JavaScript/0200_0299/0283_MoveZeroes.js
function MoveZeroesLinear(nums) {
	let length = nums.length;
	let newArray = new int[length]();
	let newArrayPointer = 0;

	for (let i = 0; i < length; i++) {
		if (nums[i] != 0) {
			newArray[newArrayPointer] = nums[i];
			newArrayPointer++;
		}
	}

	for (let i = 0; i < length; i++) {
		nums[i] = newArray[i];
	}
}

function MoveZeroesTwoPointer(nums) {
	length = nums.length;
	swapIndex = 0;

	for (i = 0; i < length; i++) {
		if (nums[i] != 0) {
			temp = nums[i];
			nums[i] = nums[swapIndex];
			nums[swapIndex] = temp;
			swapIndex++;
		}
	}
}
